add epsilon parameter to merge clusters based on distance

cmalzer opened the pull request:
Added an optional parameter epsilon, for details please see "HDBSCAN(ε^): An Alternative Cluster Extraction Method for HDBSCAN", https://arxiv.org/abs/1911.02282 
If epsilon is None or 0, results will be exactly the same as the original HDBSCAN. 
For epsilon > 0, clusters will be merged up to distance epsilon.

pep8speaks commented:
Hello @cmalzer, Thank you for submitting the Pull Request !

* In the file [`hdbscan/hdbscan_.py`](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py):

> [Line 38:63](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L38): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 46:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L46): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 48:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L48): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 52:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L52): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 52:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L52): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 52:6](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L52): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 53:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L53): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 64:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L64): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 64:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L64): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 64:16](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L64): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 66:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L66): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 124:25](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L124): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 175:25](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L175): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 211:25](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L211): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 248:25](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L248): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 314:25](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L314): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 325:101](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L325): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 358:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L358): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 358:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L358): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 359:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L359): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 361:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L361): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 634:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L634): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 634:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L634): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 634:10](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L634): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 635:1](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L635): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 635:13](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L635): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 993:18](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L993): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1008:18](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L1008): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1009:18](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L1009): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1029:72](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L1029): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1030:76](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L1030): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1031:77](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L1031): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1032:75](https://github.com/scikit-learn-contrib/hdbscan/blob/e53b1da77f37854f9bfe010c5dfe7957f894f2fe/hdbscan/hdbscan_.py#L1032): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

lmcinnes commented:
Hi -- thanks for this, it looks like a great addition to have. I only really have some minor comments. I feel like this is potentially a different cluster extraction method and that it would be useful to have documented as such. How do you feel about having ``cluster_selection_method`` being one of ``eom``, ``leaf`` or ``epsilon`` and using that fact to switch into using the merging approach. This would have the benefit of letting you set a default epsilon value (if you feel there is one). If you applied such an option then updating the docstrings accordinly would help.

Finally, do you have the time or interest to write a short tutorial about this for the documentation? Features always get more use if there is some explicit documentation in the online docs. I feel like the examples given in the paper, showing the effects, are exactly what you need -- so you are already most of the way there.

cmalzer commented:
Hi, thanks for your comments! I also thought about implementing it as new cluster_selection_method “epsilon”, but the reason why I didn’t do it is because you can combine the epsilon method with either “leaf” or “eom”. Usually there won’t be much differences between them, but there might be some if epsilon is very small. 

And yes, of course I can write a short tutorial, I will do this soon!
Regards,
Claudia

lmcinnes commented:
Ah, thanks, that makes some sense. If that's the case it may be beneficial to give epsilon a slightly more informative name, at least at the top-most user facing parts of the code. Perhaps something like ``cluster_selection_epsilon``?

cmalzer commented:
Yes, or maybe `epsilon_threshold` or `epsilon_distance_threshold `or `epsilon_cut_value`? But I don't really mind, `cluster_selection_epsilon `is also fine for me.

lmcinnes commented:
Thanks; if you want to make that change then I can either merge this with that done (and add the documentation/tutorial as a separate pull request), or I can wait for the documentation. Up to you.

cmalzer commented:
Okay, so I just changed the parameter name to cluster_selection_epsilon and I also changed the default value from None to 0.0. I included the arxiv paper as a reference (and will update it once it is published somewhere else).

I haven't written the tutorial yet; of course I can also do a separate pull request for that later.
By the way, what will be the right location for the tutorial? In docs/parameter_selection.rst?

lmcinnes commented:
You can add it to ``docs/parameter_selection.rst``, or even add a separate page for it (and add that to the ``index.rst``) if you wish. In the meantime I'm just going to merge this in. Thanks so much for this contribution -- it is greatly appreciated, and has attracted quite a bit of interest already.
